### PR TITLE
Load all persisted state on boot.

### PR DESCRIPTION
**Pull request by sgomes**
By loading all persisted state on boot, we can make adding reducers to the store a synchronous process, paving the way for dependency-based reducer injection.

#### Changes proposed in this Pull Request

* Load all persisted state on boot
* Make initial state retrieval for reducers synchronous
* Make reducer injection synchronous
* Update initial state tests

#### Testing instructions

I don't really have any specific testing instructions. This touches all of Calypso state, so take it for a spin and see if anything breaks.

### CODE REVIEW

**Comment by sgomes** [comment]
When discussing the approach implemented in this PR, the concern was raised that loading all state on boot may slow the application down. I haven't had the opportunity to compare loading performance before and after this PR, but I'll do so once I'm back from AFK.

**Comment by jsnajdr** [comment]
If I understand it correctly, this patch always stores the whole state as one IndexedDB key and removes the `storageKey` prefixing that allows to store subtrees separately?

That doesn't seem like a good idea: if we're running different entrypoints (like `login`, `signup`, `reader` and the main one) that have their own custom reducers, they will be overwriting each other's state. And the state schema is different each time.

Also, if the `domains` reducer is loaded dynamically only by `/domains` routes, and I have a saved `state.domains` state, then another Calypso session that doesn't load `/domains` will remove the stored state, because `domains` reducer is not present.

For example, going to `/posts` will have `state = { posts }` and will only store `{ posts }`, even if the persisted state was `{ domains, posts }`.

In short, the storage keys deliver a lot of flexibility and correctness and should stay. It should be easy to `loadAllState` even with them. E.g., by enumerating the IndexedDB keys that match a pattern and loading them all.

**Comment by sgomes** [comment]
> If I understand it correctly, this patch always stores the whole state as one IndexedDB key and removes the `storageKey` prefixing that allows to store subtrees separately?

@jsnajdr: No, not quite. This PR simply adds another level above all of that, simply named `calypso-state`. Everything else should remain the same underneath it.

I implemented it this way to make use of the existing functionality in `lib/browser-storage`, but you feel that it's best to maintain the existing structure instead, I could extend `lib/browser-storage` with a new `getAllStoredItems` method.

**Comment by sgomes** [comment]
@jsnajdr: Taking into account my reply above, how do you think I should proceed: add a new layer to serialised data (as is currently done in this PR), or extend `lib/browser-storage` to permit fetching all of the root-level entries through a new `getAllStoredItems` method?

**Comment by sgomes** [comment]
@jsnajdr: I went ahead and implemented `getAllStoredItems` over in #38320. If you (or any other reviewers) think that looks good and that it's best to preserve the overall structure of the current IndexedDB tree, then I'm happy to rewrite this PR on top of that.

**Comment by jsnajdr** [comment]
@sgomes I see, the patch stores one IndexedDB object named `calypso-state`, and then it has subkeys like `redux-state-12345` or `redux-state-12345:signup`.

I think we can avoid that and simply keep storing all the `redux-state-*` bits as separate keys. And the only difference will be that all keys are preloaded up-front at Calypso boot, which makes them available synchronously at any later moment.

File: client/state/initial-state.js
Comment (jsnajdr) [inline client/state/initial-state.js]: When I was testing this patch locally, my root state was never persisted to IndexedDB -- I always got only the `redux-state-1234:signup` key saved.

It's because this `map()` call has a bug: all the `persistentStoreState` calls are async tasks that run in parallel, and as they all read state from storage and write it back, they keep overwriting each other's changes:
1. Task A reads state from storage
2. Task B reads state from storage
3. Task A writes updated state to storage, but only after B's read
4. Task B writes updated state to storage, but overwrites A's changes

With the `calypso-state` umbrella key, the updates need to be orchestrated more carefully. Such bugs can be avoided by storing the `redux-state-*` bits separately, as we did until now.

Comment (sgomes) [reply client/state/initial-state.js]: Excellent catch, @jsnajdr! I'll get started on modifying this PR to work on top of #38320, then.

**Comment by sgomes** [comment]
Rebased on master to pick up node 12 changes. Will implement storage changes once  #38320 is merged.

**Comment by sgomes** [comment]
This work is currently blocked on #38320.

**Comment by sgomes** [comment]
This PR has now been rewritten on top of #38320, and no longer modifies the shape of the data being persisted to `IndexedDb`. It uses the exact same data as before, but reads all of it into memory from storage, before then synchronously initialising all reducers.

This makes the code much simpler, and should remove the concurrency issues that @jsnajdr was seeing.

I think the PR is ready for review at this point, so I'm marking it as such.

**Comment by sgomes** [comment]
Rebased on master, which removed the linter complaints.

**Comment by sgomes** [comment]
This PR does not appear to have any significant effect in startup time, with results within the same rage as before the PR.

File: client/state/initial-state.js
Comment (blowery) [inline client/state/initial-state.js]: should we worry about overlapped writes here? Imagine we call 
```
persistenStoreState( 'foo', 'foo', ..., 1 );
persistenStoreState( 'foo', 'foo', ..., 2 );
```

Can the first await return after the second await? What impact will it have?

File: client/state/initial-state.js
Comment (blowery) [inline client/state/initial-state.js]: maybe `forEach` instead of `map` since we don't need to build a return array?

File: client/state/initial-state.js
Comment (sgomes) [reply client/state/initial-state.js]: I'm not sure the onus is on the async function to handle that situation. If you're invoking an async function twice in parallel, you should be prepared for execution to happen in parallel, in any order. The correct way to invoke it twice would be:

```js
await persistentStoreState( 'foo', 'foo', ..., 1 );
await persistentStoreState( 'foo', 'foo', ..., 2 );
```

File: client/state/initial-state.js
Comment (sgomes) [reply client/state/initial-state.js]: Good point, that was a silly oversight! Switched to a for loop for clarity.

File: client/state/initial-state.js
Comment (blowery) [reply client/state/initial-state.js]: Right, but imagine two different processes are invoking it in a short time window...

Comment (sgomes) [reply client/state/initial-state.js]: My understanding is that the only thing that calls `persistentStoreState` is `persistOnChange`, and it does so with different keys, therefore causing no collisions. `persistOnChange` itself is only called once, on boot.

If we still think that there could be issues (e.g. if I'm misreading `persistOnChange` and there could indeed be key collisions), then `persistOnChange` is the right place to address them, by awaiting each write in sequence.

Comment (blowery) [reply client/state/initial-state.js]: Yeah, I'm probably overthinking this. The big case that's turning around in my head is multiple tabs writing at the same time, but even that isn't really an issue since we don't read out of storage after page load.

Comment (sgomes) [reply client/state/initial-state.js]: Ah, right. In the multiple tabs scenario, the worst that might happen is that they both get initialised from the same data, but end up wanting to write different data afterwards. I don't think this PR should have any effect on that.

The only thing that this PR tries to change is the time at which the initial data gets read; all of it will be read at the same time, now, rather than on separate occasions, immediately before the corresponding reducer gets initialised.

**Comment by gwwar** [comment]
@sgomes are we code complete here, or did we want to experiment with approaches? Happy to start trying to break the branch if we think this ready for manual testing.

**Comment by sgomes** [comment]
> @sgomes are we code complete here, or did we want to experiment with approaches? Happy to start trying to break the branch if we think this ready for manual testing.

I think we're ready for some thorough testing here, thank you!

The initial approach wasn't great, so I changed it after @jsnajdr's comments. I'm much happier with this version of the code, and I think it's ready for taking some punches. Please have at it! 😄

File: client/state/initial-state.js
Comment (gwwar) [reply client/state/initial-state.js]: > As far as I understand, the persisted data is only used as a way of showing something quickly, but it quickly gets replaced by fresh data from the API.

💯

File: client/state/initial-state.js
Comment (gwwar) [inline client/state/initial-state.js]: Do we only need this during boot? Would a WeakMap make more sense here?

**Comment by gwwar** [comment]
Light smoke testing looks good so far. 👍 I added the needs review label so E2Es could run. I'll try to poke a bit at the units tomorrow to see if we could test additional assumptions there.

File: client/state/initial-state.js
Comment (sgomes) [reply client/state/initial-state.js]: Yes, we only need this during boot, but "boot" can be kind of an odd concept when loading reducers individually. Once we modularise state, loading a particular reducer might happen at an arbitrary point in time as the user navigates through the application (e.g. only loading a stats-related reducer once the user navigates to stats).

I'm not sure that using a `WeakMap` here would translate into meaningful memory savings, but we can certainly use one defensively, as it shouldn't introduce any issues by itself 👍

Comment (sgomes) [reply client/state/initial-state.js]: Actually, now that I think about it further, I think a `WeakMap` would not work at all, precisely because we do need to preserve the state that was read in memory until that arbitrary point in time. If we really want to clear memory, we could do so by cleaning it after it's read from the cache, but I'm not sure if that's worth the risk.

**Comment by sgomes** [comment]
Rebased on master to make analysing the E2E results easier.

**Comment by kwight** [comment]
FWIW, I opened the calypso.live branch and banged away at a bunch of things: published CPTs, loaded images with the Calypso media modal in the Gutenframe editor, published changes through the iframed Customizer (`/customize`), tried an Atomic site – couldn't break anything.

File: client/state/initial-state.js
Comment (gwwar) [reply client/state/initial-state.js]: Sounds good 👍

**Review by gwwar — APPROVED** [review]
Let's give it a try! It might be worth posting for overall awareness, but let's 🚢 and see if we need to revert.

**Comment by sgomes** [comment]
Thank you, @kwight and @gwwar! I'll merge this now and stay on the lookout for any problems. A rollback shouldn't be too difficult if any issues pop up.

**Comment by sgomes** [comment]
Change deployed. I made some proactive posts to raise awareness, in case anything breaks!

**Pull request merged**
